File: hsg9-tinh/2023-hatinh-ts10/ppoint/ppoint.py
import os
import sys

input_file = os.path.join(sys.path[0], 'ppoint.inp')
output_file = os.path.join(sys.path[0], 'ppoint.out')

n = 0
P = []
min_len = sys.maxsize # độ dài nhỏ nhất cần tìm

def input_data():
    global n, P
    with open(input_file, 'r') as f:
        n = int(f.readline().strip())
        P = []
        for _ in range(n):
            x, m = map(int, f.readline().split())
            P.append((x, m))


def process():
    # Sắp xếp mảng P theo thứ tự toạ độ tăng dần
    P.sort(key=lambda point: point[0])

    # Khởi tạo toạ độ cuối cùng tìm thấy các màu
    last_green = -1
    last_red = -1
    last_yellow = -1

    global min_len

    # Duyệt mảng P
    for p in P:
        # Ghi nhận toạ độ mới khi tìm thấy màu
        if p[1] == 1:
            last_green = p[0]
        elif p[1] == 2:
            last_red = p[0]
        elif p[1] == 3:
            last_yellow = p[0]

        # Nếu đã tìm thấy đủ ba toạ độ
        # thì tính độ dài đoạn thẳng
        if last_green != -1 and last_red != -1 and last_yellow != -1:
            min_pos = min(last_green, last_red, last_yellow)
            max_pos = max(last_green, last_red, last_yellow)
            min_len = min(min_len, max_pos - min_pos)


def output():
    result = -1 if min_len == sys.maxsize else min_len
    with open(output_file, 'w') as f:
        f.write(str(result))


if __name__ == '__main__':
    input_data()
    process()
    output()

File: hsg9-tinh/2023-hatinh-ts10/ppoint/ppoint.cpp
#include <iostream>
#include <vector>
#include <algorithm>

#define inputFile "ppoint.inp"
#define outputFile "ppoint.out"

using namespace std;

struct point
{
    int x;
    int m;
};

int n;
vector<point> P;
int min_len = INT_MAX; // độ dài nhỏ nhất cần tìm

void input()
{
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);

    freopen(inputFile, "r", stdin);

    cin >> n;
    P.resize(n);

    for (int i = 0; i < n; ++i)
    {
        cin >> P[i].x >> P[i].m;
    }
}


// Hàm so sánh dùng để sắp xếp theo toạ độ
bool compare(const point& a, const point& b)
{
    return a.x < b.x;
}


void process()
{
    // Sắp xếp mảng P theo thứ tự toạ độ tăng dần
    sort(P.begin(), P.end(), compare);

    // Khởi tạo toạ độ cuối cùng tìm thấy các màu
    int last_green = -1;
    int last_red = -1;
    int last_yellow = -1;

    int min_pos;
    int max_pos;

    // Duyệt mảng P
    for (int i = 0; i < n; ++i)
    {
        // Ghi nhận toạ độ mới khi tìm thấy màu
        if (P[i].m == 1)
            last_green = P[i].x;
        else if (P[i].m == 2)
            last_red = P[i].x;
        else if (P[i].m == 3) 
            last_yellow = P[i].x;

        // Nếu đã tìm thấy đủ ba toạ độ
        // thì tính độ dài đoạn thẳng
        if (last_green != -1)
            if (last_red != -1)
                if (last_yellow != -1)
                {
                    min_pos = min({last_green, last_red, last_yellow});
                    max_pos = max({last_green, last_red, last_yellow});
                    min_len = min(min_len, max_pos - min_pos);
                }
    }
} 


void output()
{
    freopen(outputFile, "w", stdout);

    if (min_len == INT_MAX)
        cout << -1;
    else
        cout << min_len;
}


int main()
{
    input();
    process();
    output();

    return 0;
}